<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import warnings
warnings.filterwarnings("ignore")

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select
                count(*)
                from vacancies
            '''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
vacancies_total_amount = df.iloc[0,0]

print(f'Общее количество вакансий в базе: {vacancies_total_amount}')

Общее количество вакансий в базе: 49197


/tmp/ipykernel_4790/867249962.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select
                count(*)
                from employers
            '''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
employers_total_amount = df.iloc[0,0]
print(f'Общее количество работодателей в базе: {employers_total_amount}')

Общее количество работодателей в базе: 23501


/tmp/ipykernel_4790/3594696414.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select
                count(*)
                from areas
            '''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
areas_total_amount = df.iloc[0,0]
print(f'Общее количество регионов в базе: {areas_total_amount}')

Общее количество регионов в базе: 1362


/tmp/ipykernel_4790/1919600089.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select
                count(*)
                from industries
            '''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
industries_total_amount = df.iloc[0,0]
print(f'Общее количество сфер деятельности в базе: {industries_total_amount}')

Общее количество сфер деятельности в базе: 294


/tmp/ipykernel_4790/1015059550.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


***

In [16]:
# выводы по предварительному анализу данных
print(f'Общее количество вакансий в базе: {vacancies_total_amount}')
print(f'Общее количество работодателей в базе: {employers_total_amount}')
print(f'Общее количество регионов в базе: {areas_total_amount}')
print(f'Общее количество сфер деятельности в базе: {industries_total_amount}')
print()
print(f'Грубые средние оценки')
print(f'Количество вакансий на одного работодателя: {round(vacancies_total_amount/employers_total_amount)}')
print(f'Количество вакансий на один регион: {round(vacancies_total_amount/areas_total_amount)}')
print(f'Количество вакансий на одну сферу деятельности: {round(vacancies_total_amount/industries_total_amount)}')
print(f'Количество работодателей на один регион: {round(employers_total_amount/areas_total_amount)}')

Общее количество вакансий в базе: 49197
Общее количество работодателей в базе: 23501
Общее количество регионов в базе: 1362
Общее количество сфер деятельности в базе: 294

Грубые средние оценки
Количество вакансий на одного работодателя: 2
Количество вакансий на один регион: 36
Количество вакансий на одну сферу деятельности: 167
Количество работодателей на один регион: 17


**Общие выводы** по количественным данным и полученным соотношениям:<br>
*Количество вакансий больше чем работодателей в 2 раз, или 2 вакансии на одного работодателя,* это грубый вывод, требующей уточнения в дальнейшем, конечно это не означает что у каждого работодателя по 2 вакансии.<br>

Это означает что данные логичны и *потенциально* означает что у каждого работодателя есть хотя бы по одной вакансии, а может и нет, и значит необходимо *уточнить список работодателей у которых нет вакансий и в каких эти работодатели регионах и сферах деятельности.*<br>

Если таковые имеются то необходимо *учесть их долю* в общем количестве и в разрезе регионов.<br>

Судя по этим цифрам в каждом регионе по 17 работотелей, но в реальности это конечно же не так распределение работодателей по регионам *будет свидетельствовать о показателе деловой активности*, особенно если учесть уровни предлагаемых зарплат и количесвто сотрудников компании.<br>

Также интересно соотнести фактически показатель среднего числа вакансий на одного работодателя в общем к фактическому показателю среднего числа вакансий в каждом регионе. Число вакансий на одного работодателя может свидетельствовать: об уровне "текучки" кадров(хотя тут необходимо смотреть на соотношение к общему количеству сотрудников, и частоту закрытий/открытий одних и тех же вакансий на протяжении времени, т.е. добавляется временной разрез для каждой компании), развитии компании (т.е. создание новых подразделений)<br>

Если принять что в каждой сфере деятельности по 167 вакансий, то можно сказать что все они в равной мере испытывают потребность в кадрах, но это не так каждая сфера испытывает разную потребность в кадрах, и фактические значения это подтвердят. <br>

*Как отличить кадровый голод от "текучки"?* <br>

* **"Кадровый голод"** - много Новых однотипных вакансий с высокой зарплатой состоящих их однотипных должностей разных работодателей.<br>
* **"Текучка"**- постоянно закрывающиеся/открывающиеся вакансии на одну и ту же должность у одного работодателя с уровнем ЗП средний и ниже среднего по рынку.<br>

In [34]:
connection.close()

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [17]:
# текст запроса
query_4_1 = f'''select
                a.name as area,
                count(v.id) as total_vacancies
                from vacancies as v
                join areas as a on v.area_id = a.id
                group by a.name
                order by count(v.id) desc
            '''

In [18]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
display(df)

/tmp/ipykernel_5504/234815181.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area  total_vacancies
0             Москва             5333
1    Санкт-Петербург             2851
2              Минск             2112
3        Новосибирск             2006
4             Алматы             1892
..               ...              ...
764       Тарко-Сале                1
765    Новоаннинский                1
766            Бирск                1
767           Сасово                1
768             Анна                1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [6]:
# текст запроса
query_4_2 = f'''select
                    count(*)
                from vacancies
                where salary_from is not null
                    or salary_to is not null
            '''

In [7]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
not_null_salary_amount = df.iloc[0,0]
print(f'Количество вакансий у которых заполнено хотя бы одно из двух полей с зарплатой: {not_null_salary_amount}')

/tmp/ipykernel_7280/498908636.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество вакансий у которых заполнено хотя бы одно из двух полей с зарплатой: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
# текст запроса
query_4_3 = f'''select
                    avg(salary_from) as average_low_salary,
                    avg(salary_to) as average_high_salary
                from vacancies
            '''

In [9]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
average_low_salary = round(df.iloc[0,0])
average_high_salary = round(df.iloc[0,1])
print(f'Среднее значение нижней границы зарплатной вилки: {average_low_salary}')
print(f'Среднее значение верхней границы зарплатной вилки: {average_high_salary}')

Среднее значение нижней границы зарплатной вилки: 71065
Среднее значение верхней границы зарплатной вилки: 110537


/tmp/ipykernel_7280/1340936735.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [17]:
# текст запроса
query_4_4 = f'''select
                    schedule schedule_type,
                    employment employment_type,
                    count(id) total_count    
                from vacancies 
                group by schedule,employment
                order by count(id) desc    
            '''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df)

/tmp/ipykernel_4894/3274844153.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule_type      employment_type  total_count
0        Полный день     Полная занятость        35367
1   Удаленная работа     Полная занятость         7802
2      Гибкий график     Полная занятость         1593
3   Удаленная работа  Частичная занятость         1312
4     Сменный график     Полная занятость          940
5        Полный день           Стажировка          569
6     Вахтовый метод     Полная занятость          367
7        Полный день  Частичная занятость          347
8      Гибкий график  Частичная занятость          312
9        Полный день     Проектная работа          141
10  Удаленная работа     Проектная работа          133
11     Гибкий график           Стажировка          116
12    Сменный график  Частичная занятость          101
13  Удаленная работа           Стажировка           64
14     Гибкий график     Проектная работа           18
15    Сменный график           Стажировка           12
16    Вахтовый метод     Проектная работа            2
17    Сменный график     Проектная работа            1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [16]:
# текст запроса
query_4_5 = f'''select
                    experience,
                    count(id) total_count   
                from vacancies 
                group by experience
                order by count(id) asc
            '''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
display(df)

/tmp/ipykernel_4894/969833449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  total_count
0         Более 6 лет         1337
1           Нет опыта         7197
2       От 3 до 6 лет        14511
3  От 1 года до 3 лет        26152

***

In [6]:
# выводы по детальному анализу вакансий

connection.close()

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [12]:
# текст запроса
query_5_1 = f'''select
                    e.name as employer,
                    count(v.id) as total_vacancies
                from vacancies as v
                    join employers as e on v.employer_id = e.id
                group by e.name
                order by count(v.id) desc
            '''

In [5]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
display(df_5_1)

/tmp/ipykernel_10085/3282492979.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


employer  total_vacancies
0                      Яндекс             1933
1                  Ростелеком              491
2                    Тинькофф              444
3                        СБЕР              428
4               Газпром нефть              331
...                       ...              ...
14761            им. Тьюринга                1
14762        LLC Novalab Tech                1
14763   Компания Мир упаковки                1
14764  Лифинцев Олег Иванович                1
14765      ТрансЭнергоХолдинг                1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [1]:
# текст запроса
query_5_2 = f'''select
                    a.name area,
                    count(e.id) employers_count,
                    count(v.id) vac_count
                from  areas a
                    left join employers e on a.id = e.area
                    left join vacancies v on a.id = v.area_id 
                where v.name is null
                group by a.id
                
                order by 2 desc
            '''

In [6]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
display(df_5_2)


/tmp/ipykernel_21963/2075524651.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


area  employers_count  vac_count
0                Россия              410          0
1             Казахстан              207          0
2    Московская область               75          0
3    Краснодарский край               19          0
4    Ростовская область               18          0
..                  ...              ...        ...
588             Скидель                0          0
589          Белоозерск                0          0
590           Орловский                0          0
591          Павловская                0          0
592             Таксимо                0          0

[593 rows x 3 columns]

In [7]:
display(f'Регион где нет вакансий и самое болшое количество работодателей:{df_5_2.iloc[0,0]}')

'Регион где нет вакансий и самое болшое количество работодателей:Россия'

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [8]:
# текст запроса
query_5_3 = f'''select
                    e.name employer,
                    count(distinct v.area_id) areas_count    
                from  employers e 
                    left join vacancies v on e.id = v.employer_id 
                
                group by e.id
                
                order by 2 desc
            '''

In [9]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
display(df_5_3)


/tmp/ipykernel_21963/536045289.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


employer  areas_count
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
23496              Но Смокинг            0
23497        МАСТЕРСКАЯ CREED            0
23498           Dzing Finance            0
23499                    КОРС            0
23500             Наука-Связь            0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [10]:
# текст запроса
query_5_4 = f'''select
                    count(distinct e.id)    
                from  employers e 
                    left join employers_industries ei on e.id = ei.employer_id 
                where ei.industry_id is null
             '''

In [12]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
display(f'Количество работодателей у которых не указана сфера деятульности:{df_5_4.iloc[0,0]}')

/tmp/ipykernel_21963/639585724.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


'Количество  работодателей у которых не указана сфера деяетльности:8419'

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [18]:
# текст запроса
query_5_5 = f'''select
                    e.name    
                from employers_industries ei 
                    join employers e on ei.employer_id = e.id 
                group by e.name
                having count(ei.industry_id) = 4
                order by e.name
                offset 2 limit 1
             '''

In [20]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
display(f'Название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности:{df_5_5.iloc[0,0]}')


/tmp/ipykernel_21963/355499761.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


'Название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности:2ГИС'

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [33]:
# текст запроса
query_5_6 = f'''select
                    count(ei.employer_id)    
                from employers_industries ei 
                    join industries i on ei.industry_id = i.id 
                where i.name like 'Разработка программного обеспечения'
                
             '''

In [35]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
display(f'Количество работодателей у которых в качестве сферы деятельности указана "Разработка программного обеспечения":{df_5_6.iloc[0,0]}')

/tmp/ipykernel_21963/727767821.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


'Количество работодателей у которых в качестве сферы деятельности указана "Разработка программного обеспечения":3553'

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [24]:
# код для получения списка городов-милионников
tables = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')

million_cities = tuple(tables[1]['Город'])

In [20]:
# текст запроса
query_5_7 = f'''select
                    a.name "регион",
                    count(v.id) "количество вакансий"
                       
                from vacancies v 
                    join employers e on v.employer_id = e.id
                    join areas a on v.area_id = a.id
                where e.name like 'Яндекс'
                group by a.id
                having a.name in {million_cities}
                
                union all
                
                select
                    'Total',
                    count(v.id)
                        
                from vacancies v 
                    join employers e on v.employer_id = e.id
                    join areas a on v.area_id = a.id
                where e.name like 'Яндекс'
                    and a.name in {million_cities}
                    
                order by "количество вакансий"
                
             '''

In [23]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
display(df_5_7)

/tmp/ipykernel_10120/796585636.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


регион  количество вакансий
0              Омск                   21
1         Челябинск                   22
2        Красноярск                   23
3         Волгоград                   24
4             Пермь                   25
5            Казань                   25
6    Ростов-на-Дону                   25
7            Самара                   26
8               Уфа                   26
9         Краснодар                   30
10          Воронеж                   32
11      Новосибирск                   35
12  Нижний Новгород                   36
13     Екатеринбург                   39
14  Санкт-Петербург                   42
15           Москва                   54
16            Total                  485

***

In [ ]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
# текст запроса
query_6_1 = f'''select
                    count(id)    
                from vacancies    
                where name ilike '%data%' or name ilike '%данн%'
                
             '''

In [5]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
display(f'{df_6_1.iloc[0,0]} вакансий имеют отношение к данным')

'1771 вакансий имеют отношение к данным'

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [8]:
# текст запроса
query_6_2 = f'''select
                    count(id)    
                from vacancies    
                where ((name ilike '%data scientist%')
                    or (name ilike '%data science%')
                    or (name ilike '%исследователь данных%')
                    or ((name like '%ML%') and (name not like '%HTML%'))
                    or (name ilike '%machine learning%')
                    or (name ilike '%машинн%обучен%'))
                    and ((name ilike '%junior%')
                    or (experience like 'Нет опыта')
                    or (employment like 'Стажировка'))
                    
                
             '''

In [9]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
display(f'{df_6_2.iloc[0,0]} вакансий для начинающего дата-сайентиста')

'51 вакансий для начинающего дата-сайентиста'

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [33]:
# текст запроса
query_6_3 = f'''select
                    count(id)    
                from vacancies    
                where (name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or (name like '%ML%' and name not like '%HTML%')
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%')
                    and key_skills is not null
                    and (key_skills ilike '%sql%'
                    or key_skills ilike '%postgres%')
                    
                    
                
             '''

In [34]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
display(f'{df_6_3.iloc[0,0]} вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres')

'201 вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres'

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [38]:
# текст запроса
query_6_4 = f'''select
                    count(id)    
                from vacancies    
                where (name ilike '%data scientist%'
                    or name ilike '%data science%'
                    or name ilike '%исследователь данных%'
                    or (name like '%ML%' and name not like '%HTML%')
                    or name ilike '%machine learning%'
                    or name ilike '%машинн%обучен%')
                    and key_skills is not null
                    and key_skills ilike '%python%'
             '''
                    

In [39]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
display(f'{df_6_4.iloc[0,0]} вакансий для DS, в которых в качестве ключевого навыка указан Python')

'351 вакансий для DS, в которых в качестве ключевого навыка указан Python'

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса

In [ ]:
# результат запроса

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса

In [ ]:
# результат запроса

***

In [ ]:
# выводы по предметному анализу

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования